src/main/java/org/eclipsefoundation/marketplace/model/QueryParams.java

```diff
@@ -0,0 +1,142 @@
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+import org.apache.commons.lang3.StringUtils;
+
+/**
+ * Wrapper class for query parameter functionality, wrapping a Map of String to
+ * collection of String values. This class should be used anywhere where query
+ * parameters are used to control access and prevent bad data from being
+ * entered.
+ * 
+ * @author Martin Lowe
+ *
+ */
+public class QueryParams {
+	private static final String EMPTY_KEY_MESSAGE = "Key must not be null or blank";
+
+	private Map<String, List<String>> params;
+
+	/**
+	 * Builds the internal map using a copy of the data of the passed parameter map.
+	 * The map must exist to be properly set.
+	 * 
+	 * @param params list of parameters to use in the query map.
+	 */
+	public QueryParams(Map<String, List<String>> params) {
+		Objects.requireNonNull(params);
+
+		this.params = new HashMap<>(params);
+	}
+
+	/**
+	 * Retrieves the first value set in a list from the map for a given key.
+	 * 
+	 * @param params the parameter map containing the value
+	 * @param key    the key to retrieve the value for
+	 * @return the first value set in the parameter map for the given key, or null
+	 *         if absent.
+	 */
+	public String getFirstParam(String key) {
+		if (StringUtils.isBlank(key)) {
+			throw new IllegalArgumentException(EMPTY_KEY_MESSAGE);
+		}
+
+		List<String> vals = this.params.get(key);
+		if (vals == null || vals.isEmpty()) {
+			return null;
+		}
+		return vals.get(0);
+	}
+
+	/**
+	 * Retrieves the value list from the map for a given key.
+	 * 
+	 * @param params the parameter map containing the values
+	 * @param key    the key to retrieve the values for
+	 * @return the value list for the given key if it exists, or an empty collection
+	 *         if none exists.
+	 */
+	public List<String> getParams(String key) {
+		if (StringUtils.isBlank(key)) {
+			throw new IllegalArgumentException(EMPTY_KEY_MESSAGE);
+		}
+
+		List<String> vals = this.params.get(key);
+		if (vals == null || vals.isEmpty()) {
+			return Collections.emptyList();
+		}
+		return vals;
+	}
+
+	/**
+	 * <p>
+	 * Sets the given value for the given key, overriding existing values. Uses a
+	 * small load factor for the array to keep the memory footprint smaller, but the
+	 * object churn reasonable should more items need to be added.
+	 * </p>
+	 * <p>
+	 * A null value can be passed to this method for the value, which will remove
+	 * the current list values for the key if present.
+	 * </p>
+	 * 
+	 * @param params map containing parameters to update
+	 * @param key    string key to set the value for, must not be null
+	 * @param value  the value to set for the key
+	 */
+	public void setParam(String key, String value) {
+		if (StringUtils.isBlank(key)) {
+			throw new IllegalArgumentException(EMPTY_KEY_MESSAGE);
+		}
+
+		// if null value, unset the key
```

Comment:
I'd better create a real method to unset a param and make value=null an IllegalArgumentException instead in set Param. The API would much more explicit.